### PR TITLE
[WIP][SPARK-30730][SQL] Fix wrong assumption of `DateTimeUtils.converTz`

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to use the session local time zone defined by the SQL config `spark.sql.session.timeZone` instead of system JVM time zone in `DateTimeUtils.converTz`. In the case when those time zones are different, `convertTz` assumes wrongly that its input timestamp was parsed from local timestamp string using JVM time zone. Such assumption was correct in the past when timestamp strings were parsed by `java.sql.Timestamp.valueOf`. Currently, timestamp strings are converted to timestamp value via `CAST` which uses the session local time zone.

### Why are the changes needed?
This fixes wrong behavior of the `to_utc_timestamp` and `from_utc_timestamp` function in the case when system JVM time zone is different from session time zone.

### Does this PR introduce any user-facing change?
Yes, when session local time zone is different from system JVM time zone.

### How was this patch tested?
- By existing test suites. In particular, by tests for `from_utc_timestamp` and `to_utc_timestamp`.
- Added new tests to `DateFunctionsSuite` for various JVM time zone, session time zone and to/from time zones.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala [hunk 1]
```diff
@@ -65,6 +66,8 @@ object DateTimeUtils {
     TimeZone.getTimeZone(getZoneId(timeZoneId))
   }
 
+  def localTimeZone(): TimeZone = getTimeZone(SQLConf.get.sessionLocalTimeZone)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: where do we call it?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: if it's test-only, can we put it in the test suites?

**Comment by cloud-fan** [comment]
do we need to fix 2.4 as well?

**Comment by MaxGekk** [comment]
> do we need to fix 2.4 as well?

Highly likely, yes.

**Comment by MaxGekk** [comment]
I have add `[WIP]` because I have found an issue in the changes.

**Pull request closed without merging**